File: android/src/main/kotlin/com/cmc/c_shield_sdk/bridges/AipBridge.kt
package com.cmc.c_shield_sdk.bridges

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.cmc.c_shield_sdk.CShieldErrorCode
import com.example.c_shield_sdk.aip.api.AIPCore
import com.example.c_shield_sdk.utils.CShieldException
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.security.MessageDigest

class AipBridge(private val context: Context) {

    private val mainHandler = Handler(Looper.getMainLooper())

    fun handle(call: MethodCall, result: Result) {
        when (call.method) {
            // Standalone
            "aip.sign"          -> handleSign(call, result)
            "aip.verify"        -> handleVerify(call, result)
            "aip.normalizeBody" -> handleNormalizeBody(call, result)
            // Interceptor-style
            "aip.signRequest"   -> handleSignRequest(call, result)
            "aip.verifyResponse" -> handleVerifyResponse(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleSign(call: MethodCall, result: Result) {
        val payload = call.argument<String>("payload")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "payload required", null)

        Thread {
            try {
                val signature = AIPCore.sign(context, payload)
                mainHandler.post { result.success(signature) }
            } catch (e: CShieldException) {
                mainHandler.post { result.error(CShieldErrorCode.AIP_PROXY_CA, e.message, null) }
            } catch (e: Exception) {
                mainHandler.post { result.error(CShieldErrorCode.AIP_SIGNING_FAILED, e.message, null) }
            }
        }.start()
    }

    private fun handleVerify(call: MethodCall, result: Result) {
        val payload = call.argument<String>("payload")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "payload required", null)
        val signature = call.argument<String>("signature")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "signature required", null)

        Thread {
            try {
                val valid = AIPCore.verifySign(context, payload, signature)
                mainHandler.post {
                    if (valid) result.success(null)
                    else result.error(CShieldErrorCode.AIP_INVALID_SIGNATURE, "Signature verification failed", null)
                }
            } catch (e: CShieldException) {
                mainHandler.post { result.error(CShieldErrorCode.AIP_PROXY_CA, e.message, null) }
            } catch (e: Exception) {
                mainHandler.post { result.error(CShieldErrorCode.NATIVE_ERROR, e.message, null) }
            }
        }.start()
    }

    private fun handleSignRequest(call: MethodCall, result: Result) {
        val method = call.argument<String>("method")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "method required", null)
        val path = call.argument<String>("path")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "path required", null)
        val body = call.argument<ByteArray>("body") ?: ByteArray(0)
        val contentType = call.argument<String>("contentType") ?: "application/json"

        Thread {
            try {
                val request = Request.Builder()
                    .url("https://placeholder.com$path")
                    .method(method, body.toRequestBody(contentType.toMediaTypeOrNull()))
                    .build()
                val bodyResult = AIPCore.normalizeBodyForSigning(request)
                val timestamp = System.currentTimeMillis() / 1000
                val payload = "$method.$path.$timestamp.${bodyResult.hash}"
                val signature = AIPCore.sign(context, payload)
                mainHandler.post {
                    result.success(
                        mapOf(
                            "cs-timestamp" to timestamp.toString(),
                            "cs-signature" to signature,
                        )
                    )
                }
            } catch (e: CShieldException) {
                mainHandler.post { result.error(CShieldErrorCode.AIP_PROXY_CA, e.message, null) }
            } catch (e: Exception) {
                mainHandler.post { result.error(CShieldErrorCode.AIP_SIGNING_FAILED, e.message, null) }
            }
        }.start()
    }

    private fun handleVerifyResponse(call: MethodCall, result: Result) {
        val statusCode = call.argument<Int>("statusCode")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "statusCode required", null)
        val path = call.argument<String>("path")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "path required", null)
        val headers = call.argument<Map<String, String>>("headers") ?: emptyMap()
        val body = call.argument<ByteArray>("body") ?: ByteArray(0)

        val timestamp = headers["cs-timestamp"]
            ?: return result.error(CShieldErrorCode.AIP_MISSING_HEADER, "cs-timestamp missing", null)
        val signature = headers["cs-signature"]
            ?: return result.error(CShieldErrorCode.AIP_MISSING_HEADER, "cs-signature missing", null)

        Thread {
            try {
                val bodyHash = body.sha256Hex()
                val payload = "$statusCode.$path.$timestamp.$bodyHash"
                val valid = AIPCore.verifySign(context, payload, signature)
                mainHandler.post {
                    if (valid) result.success(null)
                    else result.error(CShieldErrorCode.AIP_INVALID_SIGNATURE, "Response signature verification failed", null)
                }
            } catch (e: CShieldException) {
                mainHandler.post { result.error(CShieldErrorCode.AIP_PROXY_CA, e.message, null) }
            } catch (e: Exception) {
                mainHandler.post { result.error(CShieldErrorCode.NATIVE_ERROR, e.message, null) }
            }
        }.start()
    }

    private fun handleNormalizeBody(call: MethodCall, result: Result) {
        val contentType = call.argument<String>("contentType") ?: "application/json"
        val body = call.argument<ByteArray>("body") ?: ByteArray(0)

        Thread {
            try {
                val request = Request.Builder()
                    .url("https://placeholder.com/")
                    .post(body.toRequestBody(contentType.toMediaTypeOrNull()))
                    .build()
                val bodyResult = AIPCore.normalizeBodyForSigning(request)
                mainHandler.post {
                    result.success(
                        mapOf(
                            "normalizedString" to bodyResult.normalizedString,
                            "sizeInBytes" to bodyResult.sizeInBytes,
                            "hash" to bodyResult.hash,
                        )
                    )
                }
            } catch (e: Exception) {
                mainHandler.post { result.error(CShieldErrorCode.NATIVE_ERROR, e.message, null) }
            }
        }.start()
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private fun ByteArray.sha256Hex(): String {
        val digest = MessageDigest.getInstance("SHA-256")
        return digest.digest(this).joinToString("") { "%02x".format(it) }
    }
}

File: ios/Classes/bridges/AipBridge.swift
import Flutter
import CShieldSDK

class AipBridge {

    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {

        // ── Standalone sign/verify ──────────────────────────────────────────
        case "aip.sign":          sign(args: args, result: result)
        case "aip.verify":        verify(args: args, result: result)
        case "aip.normalizeBody": normalizeBody(args: args, result: result)

        // ── Interceptor-style helpers ───────────────────────────────────────
        case "aip.signRequest":   signRequest(args: args, result: result)
        case "aip.verifyResponse": verifyResponse(args: args, result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Standalone

    private func sign(args: [String: Any], result: FlutterResult) {
        guard let payload = args["payload"] as? String else { result(invalidArg()); return }
        do {
            let signature = try AIPCore.sign(payload)
            result(signature)
        } catch {
            result(mapError(error))
        }
    }

    private func verify(args: [String: Any], result: FlutterResult) {
        guard let payload   = args["payload"] as? String,
              let signature = args["signature"] as? String else { result(invalidArg()); return }
        do {
            let valid = try AIPCore.verifySign(payload, signature: signature)
            if valid {
                result(nil)
            } else {
                result(FlutterError(code: CShieldErrorCode.aipInvalidSignature,
                                    message: "Signature verification failed", details: nil))
            }
        } catch {
            result(mapError(error))
        }
    }

    private func normalizeBody(args: [String: Any], result: FlutterResult) {
        guard let bodyData    = args["body"] as? FlutterStandardTypedData,
              let contentType = args["contentType"] as? String else { result(invalidArg()); return }

        var req = URLRequest(url: URL(string: "https://placeholder.com/")!)
        req.httpBody = bodyData.data
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")

        let bodyResult = AIPCore.normalizeBodyForSigning(req)
        result([
            "normalizedString": bodyResult.normalizedString,
            "sizeInBytes":      bodyResult.sizeInBytes,
            "hash":             bodyResult.hash,
        ])
    }

    // MARK: - Interceptor-style

    private func signRequest(args: [String: Any], result: FlutterResult) {
        guard let method      = args["method"] as? String,
              let path        = args["path"] as? String,
              let headers     = args["headers"] as? [String: String],
              let bodyData    = args["body"] as? FlutterStandardTypedData,
              let contentType = args["contentType"] as? String else { result(invalidArg()); return }

        var comps = URLComponents()
        comps.scheme = "https"; comps.host = "placeholder"
        comps.path = path.hasPrefix("/") ? path : "/\(path)"
        guard let url = comps.url else { result(invalidArg()); return }

        var req = URLRequest(url: url)
        req.httpMethod = method
        req.httpBody = bodyData.data
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")
        headers.forEach { req.setValue($1, forHTTPHeaderField: $0) }

        do {
            let interceptor = CShieldInterceptor()
            try interceptor.intercept(request: &req)
            let allHeaders = req.allHTTPHeaderFields ?? [:]
            let timestamp  = allHeaders.first { $0.key.lowercased() == "cs-timestamp" }?.value ?? ""
            let signature  = allHeaders.first { $0.key.lowercased() == "cs-signature" }?.value ?? ""
            result(["cs-timestamp": timestamp, "cs-signature": signature])
        } catch {
            result(mapError(error))
        }
    }

    private func verifyResponse(args: [String: Any], result: FlutterResult) {
        guard let statusCode = args["statusCode"] as? Int,
              let path       = args["path"] as? String,
              let headers    = args["headers"] as? [String: String],
              let bodyData   = args["body"] as? FlutterStandardTypedData else { result(invalidArg()); return }

        var comps = URLComponents()
        comps.scheme = "https"; comps.host = "placeholder"
        comps.path = path.hasPrefix("/") ? path : "/\(path)"
        guard let url = comps.url,
              let httpResp = HTTPURLResponse(url: url, statusCode: statusCode,
                                             httpVersion: "HTTP/1.1", headerFields: headers)
        else { result(invalidArg()); return }

        do {
            let interceptor = CShieldInterceptor()
            _ = try interceptor.interceptResponse(response: httpResp, data: bodyData.data)
            result(nil)
        } catch {
            result(mapError(error))
        }
    }

    // MARK: - Helpers

    private func invalidArg() -> FlutterError {
        FlutterError(code: CShieldErrorCode.invalidArgument, message: "Missing or invalid arguments", details: nil)
    }

    private func mapError(_ error: Error) -> FlutterError {
        if let e = error as? CShieldError {
            switch e {
            case .aipMissingHeader(let msg):    return FlutterError(code: CShieldErrorCode.aipMissingHeader,    message: msg, details: nil)
            case .aipTimestampExpired(let msg): return FlutterError(code: CShieldErrorCode.aipTimestampExpired, message: msg, details: nil)
            case .aipInvalidSignature(let msg): return FlutterError(code: CShieldErrorCode.aipInvalidSignature, message: msg, details: nil)
            case .aipSigningFailed(let msg):    return FlutterError(code: CShieldErrorCode.aipSigningFailed,    message: msg, details: nil)
            case .aipDetectProxyCA(let msg):    return FlutterError(code: CShieldErrorCode.aipProxyCA,          message: msg, details: nil)
            case .unimplement(let msg):         return FlutterError(code: CShieldErrorCode.nativeError,         message: msg, details: nil)
            }
        }
        return FlutterError(code: CShieldErrorCode.nativeError, message: error.localizedDescription, details: nil)
    }
}
